### PR TITLE
Add jar conflicting warning as a diagnostic

### PR DESCRIPTION
**Pull request by pramodya1994**
- Fixes https://github.com/ballerina-platform/ballerina-lang/issues/36152

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/JarResolver.java [hunk 1]
```diff
@@ -155,7 +166,16 @@ private void addPlatformLibraryPaths(PackageContext packageContext,
 
                 if (existingVersion.compareTo(newVersion) >= 0) {
                     if (!newEntry.packageName().orElseThrow().startsWith(ProjectConstants.BALLERINA_ORG)) {
-                        // TODO: issue a warning. For this we need to design diagnostic reporting in JarResolver
+                        // Issue warning for conflicting jars
+                        JBallerinaBackend.JarConflict conflict = new JBallerinaBackend.JarConflict(existingEntry,
+                                newEntry, Collections.emptyList());
+                        String conflictWarning = "Detected conflicting jar files: "
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/JarResolver.java:172]: Shall we rephrase this as `detected versions v1, v2 for <groupId>/<artifactId>`. Contnuing with v2. Also we don't need to consider this as a conflict and so listing the conflicted classes are not necessary.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/JarResolver.java [hunk 2]
```diff
@@ -57,17 +62,23 @@ public class JarResolver {
...
+        this.diagnostics = new ArrayList<>();
+    }
+
+    public List<Diagnostic> diagnostics() {
+        return diagnostics;
     }
 
     // TODO These method names are too long. Refactor them soon
-    public Collection<JarLibrary> getJarFilePathsRequiredForExecution() {
+    public Collection<JarLibrary> getJarFilePathsRequiredForExecution() { // Deprecate
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/JarResolver.java:81]: We should declare a class level variable for `jarFilePathsRequiredForExecution` and return it if it is not null.

File: compiler/ballerina-tools-api/src/main/java/io/ballerina/tools/diagnostics/DefaultDiagnostic.java [hunk 3]
```diff
@@ -45,6 +47,14 @@ class DefaultDiagnostic extends Diagnostic {
         this.message = MessageFormat.format(diagnosticInfo.messageFormat(), args);
     }
 
+    public DefaultDiagnostic(DiagnosticInfo diagnosticInfo,
```
Comment (azinneera) [inline compiler/ballerina-tools-api/src/main/java/io/ballerina/tools/diagnostics/DefaultDiagnostic.java:50]: Shall we improve the PackageDiagnostic class to support null locations without touching this class?

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateExecutableTask.java [hunk 4]
```diff
@@ -102,11 +106,25 @@ public void execute(Project project) {
 
             // Print warnings for conflicted jars
             if (!jBallerinaBackend.conflictedJars().isEmpty()) {
-                out.println("\twarning: Detected conflicting jar files:");
+                out.println("\twarning: Picking the latest jar since detected conflicting jar files:");
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateExecutableTask.java:109]: This doesn't need to change. Two versions of the same library are not considered as conflicts.

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateExecutableTask.java [hunk 5]
```diff
@@ -102,11 +106,25 @@ public void execute(Project project) {
...
                 for (JBallerinaBackend.JarConflict conflict : jBallerinaBackend.conflictedJars()) {
-                    out.println(conflict.getWarning(project.buildOptions().listConflictedClasses()));
+                    out.println(conflict.getWarning(project.buildOptions().listConflictedClasses(), false));
                 }
             }
+
+            // Report diagnostics
+            boolean hasErrors = false;
+            for (Diagnostic d : jBallerinaBackend.jarResolveDiagnosticResult().diagnostics(false)) {
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateExecutableTask.java:117]: Can't we make the jarResolver diagnostic a part of jBallerinaBackend diagnostic without exposing it like this?

Comment (pramodya1994) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateExecutableTask.java:117]: You are suggesting adding these jar resolver diagnostics to the same diagnosticsResult we already have in JBallerinaBackend?

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateExecutableTask.java (hunk 4 above)
Comment (pramodya1994) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateExecutableTask.java:109]: Reverted the change

File: compiler/ballerina-tools-api/src/main/java/io/ballerina/tools/diagnostics/DefaultDiagnostic.java (hunk 3 above)
Comment (pramodya1994) [reply compiler/ballerina-tools-api/src/main/java/io/ballerina/tools/diagnostics/DefaultDiagnostic.java:50]: Removed changes were done to the `DefaultDiagnostic` class and improved `PackageDiagnostic` class.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/JarResolver.java (hunk 1 above)
Comment (pramodya1994) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/JarResolver.java:172]: Added

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateExecutableTask.java (hunk 5 above)
Comment (azinneera) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateExecutableTask.java:117]: Yes, can we accumulate it to the jBallerinbackendDiagnostics similar to how we do for the diagnostics of manifest, compilation etc?

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/JarResolver.java (hunk 2 above)
Comment (pramodya1994) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/JarResolver.java]: Added

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateExecutableTask.java (hunk 5 above)
Comment (pramodya1994) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateExecutableTask.java:117]: Since we access jar resolver diagnostics after calling `emitExecutable` method inside `emit`, we can't show those diagnostics as compiler diagnostics. If we add these jar resolver diagnostics to the same `diagnosticResult` inside `JBallerinaBackend`, then we have to filter out and show only the diagnostics we didn't show at the compiler phase.

**Comment by anupama-pathirage** [comment]
@azinneera Do we need this PR? Looks like it got automatically closed without any attention. If you really don't need this PR anymore, please remove the `stale` label from the PR. Then it won't get added to the `[Stale PR] Ballerina Stale PR Details` mail in the future.
